core/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -914,7 +925,20 @@ public void readAllowed() throws IllegalIndexShardStateException {
         }
     }
 
-    private void writeAllowed(Engine.Operation.Origin origin) throws IllegalIndexShardStateException {
+    /** Returns timestamp of last indexing operation */
+    public long getLastWriteNS() {
+        return lastWriteNS;
+    }
+
+    private void ensureWriteAllowed(Engine.Operation op) throws IllegalIndexShardStateException {
+        lastWriteNS = op.startTime();
+        if (active.getAndSet(true) == false) {
```

Comment:
can we do this once we check the write is indeed allowed? Also I think it will be clearer if we have this in a dedicated method (`markLastWrite` or something like that)